### ISSUE TITLE
Root User Gets "permission denied" After Enabling Authentication

### ISSUE THREAD

**Issue by aauren** [coreos/etcd#9008]
After creating the root user and ensuring that the root user has the root role, enabling authentication seems to block the root user from doing anything. The documentation that I followed is here: https://github.com/coreos/etcd/blob/master/Documentation/op-guide/authentication.md#enabling-authentication

So far I have found this to be reproducible 100% of the time.

It was run with the following options:
```
# ps aux | grep etcd
etcd      4634  1.8  2.9 10557740 60548 ?      Ssl  22:03   0:07 /usr/bin/etcd -name=kubernetes-controller1 -data-dir=/var/lib/etcd -cert-file=/etc/ssl/host/cert.pem -key-file=/etc/ssl/host/pkey.pem -client-cert-auth -peer-cert-file=/etc/ssl/host/cert.pem -peer-key-file=/etc/ssl/host/pkey.pem -peer-client-cert-auth -trusted-ca-file=/etc/ssl/host/issuers_cert.pem -listen-peer-urls https://192.168.1.100:2380 -listen-client-urls https://192.168.1.100:2379,https://127.0.0.1:2379 -advertise-client-urls https://kubernetes-controller1:2379 -initial-advertise-peer-urls https://kubernetes-controller1:2380 -initial-cluster kubernetes-controller1=https://kubernetes-controller1:2380,kubernetes-controller2=https://kubernetes-controller2:2380,kubernetes-controller3.=https://kubernetes-controller3:2380 -initial-cluster-token etcd-cluster-1 -initial-cluster-state new
```

See steps for reproduction below:
```
# etcd --version
etcd Version: 3.2.9
Git SHA: v3.2.9
Go Version: go1.9.1
Go OS/Arch: linux/amd64

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" user add root
Password of root:
Type password of root again for confirmation:
User root created

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" user grant-role root root
Role root is granted to user root

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" auth enable
Authentication Enabled

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" user list
Error:  etcdserver: permission denied

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" --user root user list
Password:
Error:  etcdserver: permission denied

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" --user root auth disable
Password:
Error:  etcdserver: permission denied
```

I would expect that after creating the root user and binding it to the root role, that it should be able to do anything in the cluster including list users and disable authentication as this section of the documentation describes: https://github.com/coreos/etcd/blob/master/Documentation/op-guide/authentication.md#special-users-and-roles

The only way that I've found to come back from this error is to purge all of the data from all of the servers in the cluster and start over from scratch. As nothing seems to be able to disable authentication.

**Comment by xiang90** [comment]
i did not see you authenticate yourself as root. 

/cc @mitake might take a deeper look into this one.

**Comment by mitake** [comment]
@aauren can I see the log of your cluster?

**Comment by xiang90** [comment]
@aauren 

sorry. i did not notice that you append the user at the end for the last two commands.

**Comment by aauren** [comment]
@mitake @xiang90 any thoughts or ideas on the additional information I provided?

**Comment by mitake** [comment]
@aauren sorry for my late response. The priority of user name auth and CN auth was swapped recently and it isn't backported to stable releases yet. Could you try the master branch?

**Comment by fanyeren** [comment]
I meet the same question

**Comment by aauren** [comment]
@mitake - Sorry, I was off for a couple days, so this response is late in coming, but I can confirm that what you said is true. I built etcd from the master branch and when I run it and enable authentication then the authentication priority works correctly.

Specifically the results from my original posting change to the following:
```
# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" auth enable
Authentication Enabled

# ETCDCTL_API=3 etcdctl --cert="cert.pem" --key="pkey.pem" --endpoints="https://$(hostname -f):2379" --user root user list
Password:
root
```

Which is what I would expect based on the documentation. Thanks for your help!

Are you able to tell at this point what released version of etcd will carry these changes?

**Comment by mitake** [comment]
@aauren thanks for checking. The change will be available in 3.3.0 which would be released in the near future.

@fanyeren could you also try the latest master?

**Comment by xiang90** [comment]
@jpbetz maybe we want to backport the fix to this to 3.2? it is up to you.

i am closing this out since i am pretty sure it is already fixed in both master and 3.3 release branch.
